packages/sync-actions/src/product-types-actions.js

```diff
@@ -31,189 +23,245 @@ export function actionsMapBase(diff, previous, next, config = {}) {
-  { action: 'changeLabel', key: 'label' },
-  { action: 'setInputTip', key: 'inputTip' },
-  { actionKey: 'newValue', action: 'changeInputHint', key: 'inputHint' },
-  { action: 'changeIsSearchable', key: 'isSearchable' },
-  {
-    actionKey: 'newValue',
-    action: 'changeAttributeConstraint',
-    key: 'attributeConstraint',
-  },
-]
-
-const getIsAnAttributeDefinitionBaseFieldChange = diff =>
-  diff.label ||
-  diff.inputHint ||
-  diff.inputTip ||
-  diff.attributeConstraint ||
-  diff.isSearchable
-
-function actionsMapEnums(attributeType, attributeDiff, previous, next) {
-  const addEnumActionName =
-    attributeType === 'enum' ? 'addPlainEnumValue' : 'addLocalizedEnumValue'
-  const changeEnumOrderActionName =
-    attributeType === 'enum'
-      ? 'changePlainEnumValueOrder'
-      : 'changeLocalizedEnumValueOrder'
-  const changeEnumLabelActionName =
-    attributeType === 'enum'
-      ? 'changePlainEnumValueLabel'
-      : 'changeLocalizedEnumValueLabel'
-  const buildArrayActions = createBuildArrayActions('values', {
-    [ADD_ACTIONS]: newEnum => ({
-      attributeName: next.name,
-      action: addEnumActionName,
-      value: newEnum,
-    }),
-    [CHANGE_ACTIONS]: (oldEnum, newEnum) => {
-      const oldEnumInNext = next.values.find(
-        nextEnum => nextEnum.key === oldEnum.key
-      )
-
-      // These `changeActions` would impose a nested structure among
-      // the accumulated `updateActions` generated by `buildArrayActions`
-      // In the end; we have to flatten the structure before we pass it back
-      // to the client.
-      const changeActions = []
-      if (oldEnumInNext) {
-        if (!deepEqual(oldEnum.label, oldEnumInNext.label)) {
-          changeActions.push({
-            attributeName: next.name,
-            action: changeEnumLabelActionName,
-            newValue: newEnum,
-          })
-        } else {
-          changeActions.push({
-            attributeName: next.name,
-            action: changeEnumOrderActionName,
-            value: newEnum,
-          })
+// this is nearly similar to `buildBaseAttributesActions`, however with a significant difference.
+// `buildBasAttributesActions` generates update-actions with help of `diff`,
+// which is an object consisting of flags which indicates different operations.
+// `generateBaseFieldsUpdateActions` only generate based on `previous` and `next`.
+export const generateBaseFieldsUpdateActions = (
+  previous,
+  next,
+  actionDefinition
+) => {
+  const isEmpty = value =>
+    [isNil, nextValue => nextValue === ''].some(operator =>
+      operator(typeof value === 'string' ? value.trim() : value)
+    )
+  return Object.entries(actionDefinition).reduce(
+    (nextUpdateActions, [field, actionFieldDefinition]) => {
+      if (isEmpty(previous[field]) && isEmpty(next[field]))
+        return nextUpdateActions
+      if (!isEmpty(previous[field]) && isEmpty(next[field]))
+        return [...nextUpdateActions, actionFieldDefinition]
```

Comment:
I'm not really sure what you really mean with your question @junajan,
but for `set*`, we don't need to pass a value if `previous` had a value and `next` doesn't.

so this is completely valid:

```js
{
	action: 'setInputTip',
	attributeName: 'x'
	// no `inputTip`, means that we are un-setting the value
}
```

and for `change*` update-actions, this would _break_ in a sense that the API would throw an error. The convention is that `change*` is an indication that this is a __required__ field and can not be unset.

and we would end up with following, which is a valid update-action however with an invalid value

```js
{
	action: 'changeDescription',
	attributeName: 'x'
	// no `description`, indicates that users has attempted to either pass in an empty string, null, or undefined.
	// the API would throw an error here anyways, __even__ if we were to set a `description` key with an empty string.
}
```

Reply:
> But where do you add the actual new value for that new field? 

if you are referring to `add*` update-actions, I have an `addAttributeDefinitions` update actions defined in L124

Reply:
I see it now .. With this `if (!isEmpty(previous[field]) && isEmpty(next[field]))` you are using this action only when user removed the value (it was in `prev` but is not in `next` anymore) so you send an update action without actual value which removes value also on API. 👍